src/quantuminspire/projectq/backend_qx.py

```diff
@@ -245,54 +223,56 @@ def get_probabilities(self, qureg):
 
         Raises:
             RuntimeError: If no data is available (i.e., if the circuit has
                 not been executed). Or if a qubit was supplied which was not
                 present in the circuit (might have gotten optimized away).
         """
-        if len(self._probabilities) == 0:
+        if len(self._measured_states) == 0:
             raise RuntimeError("Please, run the circuit first!")
 
-        probability_dict = defaultdict(lambda: 0)
-
-        for state in self._probabilities:
-            mapped_state = ['0'] * len(self._measured_ids)
-            for i in range(len(qureg)):
-                logical_id = qureg[i].id
-                if logical_id in self._measured_ids:
-                    index = self._measured_ids.index(i)
-                    mapped_state[index] = state[self._logical_to_physical(logical_id)]
-            probability = self._probabilities[state]
-            probability_dict["".join(mapped_state)] += probability
-
-        return dict(probability_dict)
-
-    def _calculate_probabilities(self, counts):
-        """ Determine probabilities and set single measurement to register """
-        # Determine random outcome
-        P = random.random()
-        p_sum = 0.
-        measured = ""
-        for state in counts:
-            probability = counts[state] * 1. / self._num_runs
-            state = list(reversed(state))
-            state = "".join(state)
-            p_sum += probability
-            star = ""
-            if p_sum >= P and measured == "":
-                measured = state
-                star = "*"
-            self._probabilities[state] = probability
-            if self._verbose and probability > 0:
-                print("{0} with p = {1}{2}".format(state, probability, star))
-        return measured
+        mask_bits = map(lambda qubit: self._logical_to_physical(qubit.id), qureg)
+
+        filtered_states = QIBackend._filter_histogram(self._measured_states, mask_bits)
+
+        probability_dict = {self._map_state_to_bit_string(state, qureg): probability
+                            for state, probability in filtered_states.items()}
+
+        return probability_dict
+
+    def _map_state_to_bit_string(self, state, qureg):
+        """
+
+        Args:
+            state (int): state represented as an integer number
+            qureg (list<Qubit>): list of qubits for which to extract the state bit
+
+        Returns:
+            (string): a string of '0' and '1' corresponding to the bit value in state of each Qubit in qureg
+
+        Examples:
+            state = int('0b101010', 2)
+            qureg = [Qubit(0), Qubit(1), Qubit(5)]
+            print(self._map_state_to_bit_string(state, qureg)  # prints '011'
+
+        """
+        mapped_state = ''
+
+        for i in range(len(qureg)):
+            logical_id = qureg[i].id
+            physical_id = self._logical_to_physical(logical_id)
+            if int(state) & (1 << physical_id):
```

Comment:
Any need to int() and int?

Reply:
No, but it might also be a string (since the result from the API is decoded json, and json only allows strings as keys).